service/common/src/main/java/org/apache/polaris/service/context/DefaultRealmContextResolver.java

```diff
@@ -36,21 +38,25 @@ public DefaultRealmContextResolver(RealmContextConfiguration configuration) {
   }
 
   @Override
-  public RealmContext resolveRealmContext(
+  public CompletionStage<RealmContext> resolveRealmContext(
       String requestURL, String method, String path, Function<String, String> headers) {
-    String realm = resolveRealmIdentifier(headers);
-    return () -> realm;
+    try {
+      String realm = resolveRealmIdentifier(headers);
+      return CompletableFuture.completedFuture(() -> realm);
+    } catch (Exception e) {
+      return CompletableFuture.failedFuture(e);
+    }
   }
 
   private String resolveRealmIdentifier(Function<String, String> headers) {
     String realm = headers.apply(configuration.headerName());
     if (realm != null) {
       if (!configuration.realms().contains(realm)) {
-        throw new UnresolvableRealmContextException("Unknown realm: " + realm);
+        throw new IllegalArgumentException("Unknown realm: " + realm);
       }
     } else {
       if (configuration.requireHeader()) {
-        throw new UnresolvableRealmContextException(
+        throw new IllegalArgumentException(
```

Comment:
I guess this changes HTTP responses for unknown realms from 404 to 400 :thinking: As for me, 404 seems to be more applicable. Did you have any reason in mind for changing the status code in this case?

Reply:
Not really, the HTTP response for failed realm resolution is now crafted by the `RealmContextFilter` which is mapping all errors to 404 as before.

Reply:
Oops, I missed that :facepalm: